### PR TITLE
consider all optional brackets when splitting

### LINKED ISSUE #781: Reconsider formatting that was changed by #762
#762 fixed broken code, but in many cases the old, broken code produced subjectively better formatting, especially (in my experience) around assert statements.

Let's consider whether there's a way to improve the formatting in cases that were affected by #762.

It would be helpful to add examples to this issue of formatting changes caused by #762 that may not be improvements.

### PR DESCRIPTION
**Pull request by zsol**
I'm not married to this solution, I put it together in half an hour. This will likely cause some code churn, but hopefully for the better.

TODO:
- [x] fix tests
- [ ] verify formatting changes on a (few?) large codebase(s)
- [ ] add changelog
- [ ] better commit message

Fixes #781.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
I'm running it on our code now. A couple of things I noticed:

```
-from qcore.asserts import assert_eq, assert_dict_eq, assert_is, assert_unordered_list_eq
+from qcore.asserts import (
+    assert_eq,
+    assert_dict_eq,
+    assert_is,
+    assert_unordered_list_eq,
+)
```
Here the old line was exactly 88 characters long. I'm guessing we add the parentheses, that pushes it over 88 chars, and then we split it into multiple lines. This issue is causing a lot of churn in our codebase.

```
-    assert isinstance(value, return_type), (
-        "Value is not in one of the exepected types. Type: %s. Value: %r"
-        % (type(value), value)
+    assert isinstance(
+        value, return_type
+    ), "Value is not in one of the exepected types. Type: %s. Value: %r" % (
+        type(value),
+        value,
     )
```
Here the old formatting seems better to me.

**Comment by JelleZijlstra** [comment]
This diff shows more clearly that the optional parens are pushing the line over the limit:
```
-                    for experiment, settings in exp_settings.future_experiments.items():
+                    for (
+                        experiment,
+                        settings,
+                    ) in exp_settings.future_experiments.items():
```
The old line length was again exactly 88.

Also overall impact:
```
1559 files reformatted, 8134 files left unchanged.
```

File: tests/data/composition.py [hunk 1]
```diff
@@ -31,13 +31,15 @@ def test(self) -> None:
...
                     ):
                         print(i)
-        xxxxxxxxxxxxxxxx = Yyyy2YyyyyYyyyyy(
-            push_manager=context.request.resource_manager,
-            max_items_to_push=num_items,
-            batch_size=Yyyy2YyyyYyyyyYyyy.FULL_SIZE,
-        ).push(
-            # Only send the first n items.
-            items=items[:num_items]
+        xxxxxxxxxxxxxxxx = (
```
Comment (zsol) [inline tests/data/composition.py]: I'm not sure about this one. I slightly prefer the old one

File: tests/data/composition.py [hunk 2]
```diff
@@ -154,26 +166,30 @@ def tricky_asserts(self) -> None:
...
                       2 LOAD_CONST               1 (1)
                       4 COMPARE_OP               2 (==)
                       6 LOAD_FAST                0 (self)
                       8 STORE_ATTR               0 (x)
                      10 LOAD_CONST               0 (None)
                      12 RETURN_VALUE
-        """ % (
-            _C.__init__.__code__.co_firstlineno + 1,
+        """
+            % (_C.__init__.__code__.co_firstlineno + 1,)
```
Comment (zsol) [inline tests/data/composition.py]: This seems wrong. Haven't figured out what causes this yet

File: tests/data/expression.py [hunk 3]
```diff
@@ -491,7 +492,7 @@ def gen():
     addr_proto,
     addr_canonname,
     addr_sockaddr,
-) in socket.getaddrinfo("google.com", "http"):
+) in (socket.getaddrinfo("google.com", "http")):
```
Comment (zsol) [inline tests/data/expression.py]: superfluous parens should be stripped here

**Comment by zsol** [comment]
I updated this PR with another iteration on this idea, which hopefully causes significantly less churn, and gets some pathological cases right. There are still two bugs to figure out though.

File: black.py [hunk 4]
```diff
@@ -1363,6 +1363,8 @@ def __str__(self) -> str:
         first = next(leaves)
         res = f"{first.prefix}{indent}{first.value}"
         for leaf in leaves:
+            if leaf.type == token.LPAR and not leaf.value and leaf.parent:
+                maybe_make_parens_invisible_in_atom(leaf.parent)
```
Comment (zsol) [inline black.py]: that's a noop. `leaf` is already invisible here

File: tests/data/composition.py (hunk 1 above)
Comment (JelleZijlstra) [reply tests/data/composition.py]: Yes, the extra set of parentheses doesn't really help anything here.

File: tests/data/expression.py [hunk 5]
```diff
@@ -285,10 +285,11 @@ def gen():
 lambda a, b, c=True, *, d=(1 << v2), e="str": a
 lambda a, b, c=True, *vararg, d=(v1 << 2), e="str", **kwargs: a + b
 manylambdas = lambda x=lambda y=lambda z=1: z: y(): x()
-foo = lambda port_id, ignore_missing: {
-    "port1": port1_resource,
-    "port2": port2_resource,
-}[port_id]
+foo = (
```
Comment (JelleZijlstra) [inline tests/data/expression.py]: I think the old code is a lot nicer here.

**Comment by JelleZijlstra** [comment]
I tried it out again on our codebase. Here's a few notes:

```
                             assert (
                                 component[0] in VALID_DIRECTIVES
-                            ), "invalid directive %r" % (component,)
+                            ), ("invalid directive %r" % (component,))
```
These are just useless parens.

```
-        ] = self.detector.get_busy_time_proportion(None, time_us)
+        ] = (self.detector.get_busy_time_proportion(None, time_us))
```
Same here.


```
-        if not set(column_renames.keys()).issubset(columns_to_drop) or not set(
-            column_renames.values()
-        ).issubset(columns_to_add):
+        if (
+            not set(column_renames.keys()).issubset(columns_to_drop)
+            or not set(column_renames.values()).issubset(columns_to_add)
+        ):
```
This is great.

```
-            z += h.Button(
-                class_=["ui", "button"],
-                id="clear-search-button",
-                onclick="resetFilter()",
-            )(h.text("Clear"))
+            z += (
+                h.Button(
+                    class_=["ui", "button"],
+                    id="clear-search-button",
+                    onclick="resetFilter()",
+                )(h.text("Clear"))
+            )
```
Another unnecessary set of parentheses added.

**Comment by JelleZijlstra** [comment]
Also in terms of churn the effect is similar:
`2289 files reformatted, 10702 files left unchanged, 6 files failed to reformat.`

The crashes are due to instability. IThe first pass is:

```             data[
                 "share_data"
-            ] = yield a.app.view.share.mobile.get_native_share_action.asynq(share)
+            ] = (yield a.app.view.share.mobile.get_native_share_action.asynq(share))
```
And the second is:
```
-            data[
-                "share_data"
-            ] = (yield a.app.view.share.mobile.get_native_share_action.asynq(share))
+            data["share_data"] = (
+                yield a.app.view.share.mobile.get_native_share_action.asynq(share)
+            )
```
The eventual result is an improvement though. The other crashes are also cases where the first pass adds parentheses, and the second one rearranges the structure of an assignment + yield.

**Comment by zsol** [comment]
This latest one seems promising. I've run it on a large codebase and it did produce a bunch of changes, but all of the changes look better than without this PR.

Most of the changes follow this pattern:
```
-        if not self.ignore_interrupt and isinstance(
-            sys.exc_info()[1], KeyboardInterrupt
+        if (
+            not self.ignore_interrupt
+            and isinstance(sys.exc_info()[1], KeyboardInterrupt)
         ):
```

Some asserts have changed as well:

```
-        assert isinstance(
-            types, collections.Iterable
-        ), "types must be iterable, is %r" % (types,)
+        assert isinstance(types, collections.Iterable), (
+            "types must be iterable, is %r" % (types,)
+        )
```

The most questionable changes are around manually split strings:
```
-        sql = "INSERT INTO status_type (type_name, seq)  " "VALUES ('{}', {})".format(
-            key, val
+        sql = (
+            "INSERT INTO status_type (type_name, seq)  "
+            "VALUES ('{}', {})".format(key, val)
         )
```

All in all, I think this PR makes the formatting better, so ready for review.

**Comment by zsol** [comment]
As for the amount of churn: on a codebase with half a million LOC across 3k+ files this PR generates
```
41 files changed, 188 insertions(+), 151 deletions(-)
```

**Comment by JelleZijlstra** [comment]
I ran it on our codebase again. This one is a little questionable:
```
     def table_string(self):
-        return "%5d  %5d  %5d  %5d  %5d  %5d  %5d  %5d  %5d  %5d" % (
-            self.task_count,
-            self.delete_count,
-            self.suspend_count,
-            self.call_count,
-            self.send_count,
-            self.throw_count,
-            self.return_count,
-            self.cancel_count,
-            self.pause_count,
-            self.resume_count,
+        return (
+            "%5d  %5d  %5d  %5d  %5d  %5d  %5d  %5d  %5d  %5d"
+            % (
+                self.task_count,
+                self.delete_count,
+                self.suspend_count,
+                self.call_count,
+                self.send_count,
+                self.throw_count,
+                self.return_count,
+                self.cancel_count,
+                self.pause_count,
+                self.resume_count,
+            )
         )
```
On the other hand there are similar cases to this where the tuple after % fits on one line and it looks nicer, so maybe this is the price we have to pay.

This still inserts unnecessary parens:
```
         assert email_type not in {
             aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
             aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
             aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
-        }, "should not have xxxxxxxxxxxxxxxxxxxxxxx %s" % (email_type)
+        }, ("should not have xxxxxxxxxxxxxxxxxxxxxxx %s" % (email_type))
```

Overall impact: `604 files reformatted, 12022 files left unchanged.`

I agree that most of the changes now are improvements, so I'd be OK with applying it (except for the unnecessary parens in the second diff snippet above).

**Comment by zsol** [comment]
I tried hard to remove the unnecessary parens you highlight, but I realized this is a bug that's better addressed in a separate issue. I'm reasonably sure the root cause is the same as with this one:

Black on master does this:
```
-        assert looooooooooooooooooooong in { looooooooooooooooooooooooooooooooooooooooooooooooooooooong}, "wot. a bug?!" % loooooooong
+        assert looooooooooooooooooooong in {
+            looooooooooooooooooooooooooooooooooooooooooooooooooooooong
+        }, ("wot. a bug?!" % loooooooong)
```

**Comment by nchammas** [comment]
@zsol - I read [your comment about fluent interfaces](https://github.com/psf/black/issues/781#issuecomment-490466808) and came here to check this PR out. Could you share an example or two of how this PR might improve the formatting of fluent-style programs?

My team heavily uses Pandas, Apache Spark, and SQLAlchemy, so we are writing code in this style all the time. I would be very happy to see any improvements to how Black handles this style of code, since it would make it easier for me to sell Black to my team.

**Comment by ambv** [comment]
Rejected in favor of #1092. Maybe we'll revisit the approach in the future.

**Pull request closed without merging**